snuba/utils/kafka/consumers/abstract.py

```diff
@@ -0,0 +1,34 @@
+from abc import ABC, abstractmethod
+from typing import Callable, Generic, NewType, Optional, Sequence, Tuple, TypeVar
+
+
+TMessage = TypeVar("TMessage")
+
+Topic = NewType("Topic", str)
+Partition = NewType("Partition", int)
+Offset = NewType("Offset", int)
+
+
+class Consumer(Generic[TMessage], ABC):
+    @abstractmethod
+    def subscribe(
+        self,
+        topics: Sequence[str],
+        on_assign: Optional[Callable[[Sequence[Tuple[Topic, Partition]]], None]] = None,
```

Comment:
I think it's reasonable to make this a NamedTuple since it is going to be used pretty frequently.

> I believe the confluent api already has a type for this TopicPartition. Which includes the offset as well. What was the reason not to use it ?

I'm glad you asked this question. There are a few reasons:

1. I didn't want to introduce any direct coupling to the Confluent consumer in this abstract consumer, mostly from an ideological standpoint. In reality, I don't anticipate the driver changing, but it seems easy to keep this separate, and also makes it slightly easier to mock.
2. The Confluent driver passes around the offset in many cases where I haven't ever observed a valid offset being provided, like in the subscription callbacks. This makes it appear superficially as you can use that offset in this callback, which will lead to either incorrect behavior, or runtime errors. Since the "invalid offset" value is the same type as a valid value (`int`), you can't rely on the type checker to make it difficult to do the wrong thing.
3. I think representing the error as an attribute on the object rather than throwing an exception makes it challenging to write code that does the right thing in the face of failure (which is usually just to crash when something unexpected happens.)
4. The native TopicPartition (or an object sharing the same structure) can't easily be used as an object representing the partition (for example as a mapping key or set element) because the offset (or error) would be considered in the comparison. (Considering this, I'll also change the return type of `commit` to be `Optional[Mapping[TopicPartitionKey, Offset]]` instead.)

Reply:
_The Confluent driver passes around the offset in many cases where I haven't ever observed a valid offset being provided_

Why doesn't this surprise me ?

ok, then.